Python/optimizer_bytecodes.c

```diff
@@ -947,14 +950,54 @@ dummy_func(void) {
             // isinstance(inst, cls) where both inst and cls have
             // known types, meaning we can deduce either True or False
 
-            // The below check is equivalent to PyObject_TypeCheck(inst, cls)
             PyObject *out = Py_False;
-            if (inst_type == cls_o || PyType_IsSubtype(inst_type, cls_o)) {
+            if (sym_IS_SUBTYPE(inst_type, cls_o)) {
                 out = Py_True;
             }
             sym_set_const(res, out);
             REPLACE_OP(this_instr, _POP_CALL_TWO_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)out);
         }
+        else if (inst_type && sym_matches_type(cls, &PyTuple_Type)) {
+            // isinstance(inst, tup) where inst has a known type and tup is a tuple.
+            // We can deduce True if inst is an instance of at least one of
+            // the items in the tuple.
+            // We can deduce False if all items in the tuple have known types and
+            // inst is not an instance of any of them.
+
+            int length = sym_tuple_length(cls);
+            if (length != -1) {
+                // We cannot do anything about tuples with unknown length
+                bool all_items_known = true;
+                PyObject *out = NULL;
+                for (int i = 0; i < length; i++) {
+                    JitOptSymbol *item = sym_tuple_getitem(ctx, cls, i);
+                    if (!sym_has_type(item)) {
+                        // There is an unknown item in the tuple,
+                        // we can no longer deduce False.
+                        all_items_known = false;
+                        continue;
+                    }
```

Comment:
We can't infer *anything* in this case, since the class could do *anything* in its `__instancecheck__` or whatever, so it's no longer side-effect free. Need to bail on the whole optimization at this point.

So I don't think we need `all_items_known`, either. We can either:
- Break early on our first `True`, like you do below, and infer `True`.
- Loop over *everything* and infer `False`.
- Bail on an unknown thing and infer `bool`.

Reply:
> We can't infer anything in this case, since the class could do anything in its __instancecheck__ or whatever, so it's no longer side-effect free. Need to bail on the whole optimization at this point.

Just to check if I understood this point: even if we have something like `isinstance(42, (unknown, int))` we can't infer `True` because if `unknown` defines its own `__instancecheck__`, it would change the semantics of the program if we infer `True`. This is because `unknown.__instancecheck__` would no longer be called, right?

So basically what you said, once we see an item with an unknown type, we must stop.

Reply:
> Basically, yeah. It's a bit more subtle, though: we can infer the result to be True, but we can't remove the isinstance call itself.

Awesome, that was pretty much my understanding as well :)

I'll fix the PR hopefully this weekend.